### PR TITLE
dBFT: recalculate proposed Merkle state if needed

### LINKED ISSUE #60: Optimize the way how dBFT-accepted blocks are being inserted into chain
Currently after #49 every block built by dBFT is being routed to `blockQueue` instance that is trying to send the block back to miner (if a proper sealing task is found) or to insert it to chain directly. In case of direct chain insertion `blockQueue` instance doesn't have transaction receipts being calculated for this block. Latter occurs mostly when the node is not a primary of the current block. Receipts are being recalculated by `Blockchain` during the block insertion. However, after #53 we'll have a proper set of receipts calculated for every PrepareRequest on the backup node side (for verification purposes).

Within the scope of this issue we can store receipts calculated during PrepareRequest verification and pass the new dBFT block to the `blockQueue` instance along with the corresponding receipts. This allows to omit double receipts recalculation and save time on computations. For example of how block with receipts, state and log can be stores, refer to miner's code: https://github.com/bane-labs/go-ethereum/blob/94233fe05ce42ac2e4c4556ea1b3b871f2741964/miner/worker.go#L686

This issue is roughly depends on #53 and is not a high priority.

### LINKED ISSUE #140: Invalid merkle root of accepted block 128928 in T2
#### System information

Commit hash: originally discovered in early version deployed on T2 (81ea1cec0e3701ab9e52387be29ada71f4d5dbcf) with dBFT proposal verification disabled, but supposed to be present at the current bane-main (7b56bf71af205bb183265e49513e6226d69a9f13) and t2 branches (a4cea5bb6957b3deaa39da10e2e6736487c61dec).

Bug occured on T2 13 Feb 2024.

#### Expected behaviour

Merkle root of accepted block is always valid.

#### Actual behaviour

Somehow Merkle root of accepted block is invalid. Seed node can't import block with invalid Merkle root and is stuck at block 128927 of T2. Consensus nodes can't import block with invalid Merkle root via P2P, but can store this invalid block received via consensus engine. CNs continue new blocks acceptance afterwords.

#### Steps to reproduce the behaviour

Caught on T2 with 5 out of 7 CNs alive and single RPC (seed) node. I won't attach the full logs since GH doesn't allow large files, here are only logs around the "broken" block that are enough to recreate the picture:

[cn1_bad-block.log](https://github.com/bane-labs/go-ethereum/files/14375742/cn1_bad-block.log)
[cn2_bad-block.log](https://github.com/bane-labs/go-ethereum/files/14375748/cn2_bad-block.log)
[cn3_bad-block.log](https://github.com/bane-labs/go-ethereum/files/14375749/cn3_bad-block.log)
[cn4_bad-block.log](https://github.com/bane-labs/go-ethereum/files/14375752/cn4_bad-block.log)
[cn5_bad-block.log](https://github.com/bane-labs/go-ethereum/files/14375753/cn5_bad-block.log)
[seed1_bad-block.log](https://github.com/bane-labs/go-ethereum/files/14375839/seed1_bad-block.log)

Nodes genesis configuration file:
[genesis.json](https://github.com/bane-labs/go-ethereum/files/14375913/genesis.json)

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #140. Close #60. Port #141 with some changes.

In this PR:
* Ensure all header's fields that affect miner's state calculations are properly set in `(*DBFT).Prepare`. A part of #140.
* Recalculate Merkle state depending on updated dBFT context information in case of CV. Close #140.
* Optimize accepted block insertion into chain using state calculated during PrepareRequest-level block verification (port the #124). Close #60.

Tested on privnet with transactions, works as expected. @chenquanyu, @nicolegys, welcome to review and test.

### CODE REVIEW

File: consensus/dbft/dbft.go [hunk 1]
```diff
@@ -448,10 +423,34 @@ func New(config *params.DBFTConfig, _ ethdb.Database) (*DBFT, error) {
...
+				ethBlock := dbftBlock.ToEthBlock()
+
+				state, receipts, _, _, err := c.chain.ProcessState(ethBlock)
+				if err != nil {
+					w := bytes.NewBuffer(nil)
+					_ = ethBlock.EncodeRLP(w)
+
+					log.Crit("failed to process proposal state",
+						"err", err,
+						"rlp", hex.EncodeToString(w.Bytes()))
```
Comment (roman-khimov) [inline consensus/dbft/dbft.go]: Can be hunderds of KBs.

File: consensus/dbft/dbft.go [hunk 2]
```diff
@@ -448,10 +423,34 @@ func New(config *params.DBFTConfig, _ ethdb.Database) (*DBFT, error) {
...
+				dbftBlock := c.newBlockFromContext(c.sealingProposal)
+				dbftBlock.transactions = c.sealingTransactions
+				ethBlock := dbftBlock.ToEthBlock()
+
+				state, receipts, _, _, err := c.chain.ProcessState(ethBlock)
+				if err != nil {
+					w := bytes.NewBuffer(nil)
+					_ = ethBlock.EncodeRLP(w)
+
+					log.Crit("failed to process proposal state",
```
Comment (roman-khimov) [inline consensus/dbft/dbft.go]: "proposed state"? "state from proposal"?

File: consensus/dbft/dbft.go [hunk 3]
```diff
@@ -448,10 +423,34 @@ func New(config *params.DBFTConfig, _ ethdb.Database) (*DBFT, error) {
...
+						"err", err)
+				}
+
+				c.sealingProposal = b.Header()
+			}
+
 			// Fill in only proposal and receipts, transactions will be properly
 			// set from context later in SetTransactionHashes callback.
 			req.SealingProposal = c.sealingProposal
-
```
Comment (roman-khimov) [inline consensus/dbft/dbft.go:462]: :eyes:

File: consensus/dbft/dbft.go [hunk 4]
```diff
@@ -582,6 +581,38 @@ func New(config *params.DBFTConfig, _ ethdb.Database) (*DBFT, error) {
...
+	h := types.CopyHeader(sealingProposal)
+
+	// BlockIndex -> Number
+	h.Number = big.NewInt(int64(ctx.BlockIndex))
+
+	// PrimaryIndex -> Nonce
+	binary.BigEndian.PutUint64(h.Nonce[:], uint64(ctx.PrimaryIndex))
+
+	h.Difficulty = c.getDifficulty(int(ctx.PrimaryIndex), uint64(ctx.BlockIndex))
+
```
Comment (roman-khimov) [inline consensus/dbft/dbft.go:612]: MixDigest?

File: consensus/dbft/block.go [hunk 5]
```diff
@@ -3,6 +3,7 @@ package dbft
 import (
 	"errors"
 	"fmt"
+	"github.com/ethereum/go-ethereum/core/state"
```
Comment (roman-khimov) [inline consensus/dbft/block.go]: goimports

File: consensus/dbft/dbft.go (hunk 1 above)
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go]: Fixed, added only several fields instead.

File: consensus/dbft/dbft.go (hunk 2 above)
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go]: Fixed.

File: consensus/dbft/dbft.go (hunk 3 above)
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go:462]: Moved to a separate comment since this change is intentional.

File: consensus/dbft/dbft.go (hunk 4 above)
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go:612]: MixDigest is set during PrepareRequest construction and CNs will agree on it during consensus, so it shouldn't be here anymore. Ref. https://github.com/bane-labs/go-ethereum/pull/121#discussion_r1473416855.

File: consensus/dbft/block.go (hunk 5 above)
Comment (AnnaShaleva) [reply consensus/dbft/block.go]: Fixed.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**